pyramid_oereb/contrib/print_proxy/mapfish_print/toc_pages.py

```diff
@@ -5,116 +5,197 @@
+        page_bottom_margin = 20  # toc.jrxml
+        footer_height = 10  # toc.jrxml
+        total_size += page_bottom_margin + footer_height
+        log.debug(f"header total_size: {total_size}")
+        return total_size
+
     def compute_d1(self):
+        # The ConcernedTheme-Heading cannot be calculated at runtime. The used label
+        # is defined in the mfp-templates which are not accessible here. Therefore
+        # we assume one line without a line break.
+        log.debug(f"d1 total_size: {self.d1_height}")
         return self.d1_height
 
     def compute_d2(self):
-        x = len(self.extract['ConcernedTheme'] * self.toc_item_height)
-        if x > self.d2_height:
-            return x
+        total_size = 0
+        blank_space_above = 2  # toc.jrxml
+        page_label_height = 10  # toc.jrxml
+        total_size += blank_space_above + page_label_height
+        toc_item_height = 17  # toc.jrxml (20 in tocConcernedTheme.jrxml)
+        unique_concerned_themes = []
+        for concerned_theme in self.extract["ConcernedTheme"]:
+            if concerned_theme not in unique_concerned_themes:
+                unique_concerned_themes.append(concerned_theme)
+        total_size += len(unique_concerned_themes) * toc_item_height
+        log.debug(f"d2 total_size: {total_size}")
+        if total_size > self.d2_height:
+            return total_size
         else:
             return self.d2_height
 
     def compute_d3(self):
-        x = self.not_concerned_themes_title_height + len(self.extract['NotConcernedTheme'] * self.not_concerned_themes_item_height)  # noqa
-        if x > self.d3_height:
-            return x
+        total_size = 0
+        blank_space_above = 26  # toc.jrxml
+        not_concerned_themes_title_height = 15  # toc.jrxml
+        blank_space_between = 5  # toc.jrxml
+        not_concerned_themes_item_height = 15  # toc.jrxml (12 in themelist.jrxml)
+        total_size += (
+            blank_space_above + not_concerned_themes_title_height + blank_space_between
+        )
+        total_size += (
+            len(self.extract["NotConcernedTheme"]) * not_concerned_themes_item_height
+        )
+        log.debug(f"d3 total_size: {total_size}")
+        if total_size > self.d3_height:
+            return total_size
         else:
             return self.d3_height
 
     def compute_d4(self):
+        # The NoDataTheme-Heading cannot be calculated at runtime. The used label
+        # is defined in the mfp-templates which are not accessible here. Therefore
+        # we assume one line without a line break.
+        log.debug(f"d4 total_size: {self.d4_height}")
         return self.d4_height
 
     def compute_d5(self):
-        x = len(self.extract['ThemeWithoutData'] * self.theme_without_data_item_height)
-        if x > self.d5_height:
-            return x
+        total_size = 0
+        theme_without_data_item_height = 15  # toc.jrxml (12 in themelist.jrxml)
+        total_size += len(
```

Comment:
total_size should be calculated like this: **len(**self.extract["ThemeWithoutData"]**)** * theme_without_data_item_height
Also, the theme_without_data_item_height is 15 only for one entry. 12 is the line height of an individual entry, plus padding. (@voisardf )

Reply:
I just pushed a bugfix adressing this. Did I understand it correctly?